File: scripts/test_client.py
#!/usr/bin/env python3
"""Run Needle HTTP checks against a running service (host or remote)."""

from __future__ import annotations

import argparse
import json
import os
import sys
import urllib.error
import urllib.request


def request_json(
    base_url: str,
    method: str,
    path: str,
    body: dict | None = None,
    timeout: float = 120.0,
) -> tuple[int, object]:
    url = f"{base_url.rstrip('/')}{path}"
    data = None
    headers = {"Accept": "application/json"}
    if body is not None:
        data = json.dumps(body).encode("utf-8")
        headers["Content-Type"] = "application/json"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=timeout) as resp:
            payload = resp.read().decode("utf-8")
            return resp.status, json.loads(payload) if payload else None
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        try:
            parsed = json.loads(detail)
        except json.JSONDecodeError:
            parsed = detail
        return exc.code, parsed


def check(name: str, ok: bool, detail: str = "") -> bool:
    mark = "PASS" if ok else "FAIL"
    line = f"  [{mark}] {name}"
    if detail:
        line = f"{line} — {detail}"
    print(line)
    return ok


def main() -> int:
    parser = argparse.ArgumentParser(description="Test Needle HTTP service from outside Docker.")
    parser.add_argument(
        "--base-url",
        default=os.environ.get("NEEDLE_BASE_URL", "http://localhost:8888"),
        help="Service root URL (env: NEEDLE_BASE_URL)",
    )
    parser.add_argument("--verbose", "-v", action="store_true", help="Print full JSON responses")
    args = parser.parse_args()
    base = args.base_url

    print(f"Needle client → {base}\n")
    passed = 0
    total = 0

    def run_check(name: str, ok: bool, detail: str = "", payload: object | None = None) -> None:
        nonlocal passed, total
        total += 1
        if check(name, ok, detail):
            passed += 1
        if args.verbose and payload is not None:
            print(json.dumps(payload, indent=2))
            print()

    status, health = request_json(base, "GET", "/health")
    run_check(
        "GET /health",
        status == 200
        and isinstance(health, dict)
        and health.get("status") == "ok"
        and health.get("warmed") is True,
        f"HTTP {status}",
        health,
    )

    status, tools_payload = request_json(base, "GET", "/tools")
    tool_count = 0
    if isinstance(tools_payload, dict) and isinstance(tools_payload.get("tools"), list):
        tool_count = len(tools_payload["tools"])
    run_check(
        "GET /tools",
        status == 200 and tool_count >= 3,
        f"{tool_count} tools",
        tools_payload,
    )

    status, run_body = request_json(
        base,
        "POST",
        "/run",
        {"query": "what is the weather in Lagos?"},
    )
    results_ok = False
    if isinstance(run_body, dict):
        results = run_body.get("results")
        if isinstance(results, list) and len(results) > 0:
            first = results[0]
            if isinstance(first, dict) and first.get("city") == "Lagos":
                results_ok = True
    run_check(
        "POST /run",
        status == 200 and results_ok,
        "expected Lagos weather in results",
        run_body,
    )

    status, complete_body = request_json(
        base,
        "POST",
        "/complete",
        {"text": "set thermostat to 21 cool"},
    )
    calls_ok = False
    if isinstance(complete_body, dict):
        calls = complete_body.get("function_calls")
        if isinstance(calls, list) and len(calls) > 0:
            first = calls[0]
            if (
                isinstance(first, dict)
                and first.get("name") == "set_thermostat"
                and isinstance(first.get("arguments"), dict)
            ):
                calls_ok = True
    run_check(
        "POST /complete (server tools)",
        status == 200 and calls_ok,
        "expected set_thermostat call",
        complete_body,
    )

    client_tools = [
        {
            "name": "set_lights",
            "description": "Turn room lights on or off",
            "parameters": {
                "type": "object",
                "properties": {
                    "room": {"type": "string"},
                    "on": {"type": "boolean"},
                },
                "required": ["room", "on"],
            },
        }
    ]
    status, complete_custom = request_json(
        base,
        "POST",
        "/complete",
        {"text": "turn on the kitchen lights", "tools": client_tools},
    )
    custom_ok = False
    if isinstance(complete_custom, dict):
        calls = complete_custom.get("function_calls")
        if isinstance(calls, list) and len(calls) > 0:
            first = calls[0]
            if isinstance(first, dict) and first.get("name") == "set_lights":
                custom_ok = True
    run_check(
        "POST /complete (client tools)",
        status == 200 and custom_ok,
        "expected set_lights call",
        complete_custom,
    )

    status, extract_body = request_json(
        base,
        "POST",
        "/extract",
        {
            "text": "Invoice from Acme Corp total 1200 USD",
            "schema": {
                "type": "object",
                "properties": {
                    "vendor": {"type": "string"},
                    "total": {"type": "number"},
                },
                "required": ["vendor", "total"],
            },
        },
    )
    extract_ok = False
    if isinstance(extract_body, dict):
        result = extract_body.get("result")
        if isinstance(result, dict) and result.get("vendor") == "Acme Corp":
            extract_ok = True
    run_check(
        "POST /extract",
        status == 200 and extract_ok,
        "expected Acme Corp invoice",
        extract_body,
    )

    print(f"\n{passed}/{total} checks passed")
    return 0 if passed == total else 1


if __name__ == "__main__":
    sys.exit(main())

File: scripts/test_client.mjs
#!/usr/bin/env node
/**
 * Test Needle HTTP service from outside Docker (Node 18+).
 *
 *   NEEDLE_BASE_URL=http://localhost:8888 node scripts/test_client.mjs
 */

const baseUrl = (process.env.NEEDLE_BASE_URL || "http://localhost:8888").replace(
  /\/$/,
  "",
);

const verbose = process.argv.includes("-v") || process.argv.includes("--verbose");

async function requestJson(method, path, body) {
  const res = await fetch(`${baseUrl}${path}`, {
    method,
    headers: body ? { "Content-Type": "application/json" } : undefined,
    body: body ? JSON.stringify(body) : undefined,
  });
  const text = await res.text();
  let data = null;
  if (text) {
    data = JSON.parse(text);
  }
  return { status: res.status, data };
}

function check(name, ok, detail = "") {
  const mark = ok ? "PASS" : "FAIL";
  const suffix = detail ? ` — ${detail}` : "";
  console.log(`  [${mark}] ${name}${suffix}`);
  return ok;
}

let passed = 0;
let total = 0;

function runCheck(name, ok, detail, payload) {
  total += 1;
  if (check(name, ok, detail)) {
    passed += 1;
  }
  if (verbose && payload !== undefined) {
    console.log(JSON.stringify(payload, null, 2));
    console.log();
  }
}

console.log(`Needle client → ${baseUrl}\n`);

const health = await requestJson("GET", "/health");
runCheck(
  "GET /health",
  health.status === 200 &&
    health.data?.status === "ok" &&
    health.data?.warmed === true,
  `HTTP ${health.status}`,
  health.data,
);

const tools = await requestJson("GET", "/tools");
const toolCount = Array.isArray(tools.data?.tools) ? tools.data.tools.length : 0;
runCheck(
  "GET /tools",
  tools.status === 200 && toolCount >= 3,
  `${toolCount} tools`,
  tools.data,
);

const run = await requestJson("POST", "/run", {
  query: "what is the weather in Lagos?",
});
const runOk =
  run.status === 200 &&
  Array.isArray(run.data?.results) &&
  run.data.results[0]?.city === "Lagos";
runCheck("POST /run", runOk, "expected Lagos weather in results", run.data);

const complete = await requestJson("POST", "/complete", {
  text: "set thermostat to 21 cool",
});
const completeOk =
  complete.status === 200 &&
  complete.data?.function_calls?.[0]?.name === "set_thermostat";
runCheck(
  "POST /complete (server tools)",
  completeOk,
  "expected set_thermostat call",
  complete.data,
);

const completeCustom = await requestJson("POST", "/complete", {
  text: "turn on the kitchen lights",
  tools: [
    {
      name: "set_lights",
      description: "Turn room lights on or off",
      parameters: {
        type: "object",
        properties: {
          room: { type: "string" },
          on: { type: "boolean" },
        },
        required: ["room", "on"],
      },
    },
  ],
});
const customOk =
  completeCustom.status === 200 &&
  completeCustom.data?.function_calls?.[0]?.name === "set_lights";
runCheck(
  "POST /complete (client tools)",
  customOk,
  "expected set_lights call",
  completeCustom.data,
);

const extract = await requestJson("POST", "/extract", {
  text: "Invoice from Acme Corp total 1200 USD",
  schema: {
    type: "object",
    properties: {
      vendor: { type: "string" },
      total: { type: "number" },
    },
    required: ["vendor", "total"],
  },
});
const extractOk =
  extract.status === 200 && extract.data?.result?.vendor === "Acme Corp";
runCheck(
  "POST /extract",
  extractOk,
  "expected Acme Corp invoice",
  extract.data,
);

console.log(`\n${passed}/${total} checks passed`);
process.exit(passed === total ? 0 : 1);
